packages/eslint-plugin-template/tests/rules/no-duplicate-attributes.test.ts

```diff
@@ -0,0 +1,149 @@
+import {
+  convertAnnotatedSourceToFailureCase,
+  RuleTester,
+} from '@angular-eslint/utils';
+import rule, {
+  MessageIds,
+  RULE_NAME,
+} from '../../src/rules/no-duplicate-attributes';
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({
+  parser: '@angular-eslint/template-parser',
+});
+
+const messageId: MessageIds = 'noDuplicateAttributes';
+
+ruleTester.run(RULE_NAME, rule, {
+  valid: [
+    '<input name="foo">',
+    '<input [name]="foo">',
+    '<input (change)="bar()">',
+    '<input [(ngModel)]="foo">',
+    '<input [(ngModel)]="model" (ngModelChange)="modelChanged()">',
+    '<div (@fade.start)="animationStarted($event)" (@fade.done)="animationDone($event)"></div>',
+    '<div (window:keydown)="windowKeydown($event)" (document:keydown)="documentKeydown($event)" (document:keyup)="documentKeyup($event)" (keyup)="keyup($event)" (keydown)="keydown($event)"></div>',
+    '<div [style.width.px]="col.width" [width]="col.width"></div>',
+    '<button [class.disabled]="!enabled" [disabled]="!enabled"></button>',
```

Comment:
@rafaelss95 it doesn't look like `[style.foo]` and `[class.foo]` are represented in [the AST](https://astexplorer.net/#/gist/790639325c3eb4acf9860199d260fa02/9254a75724151f64163fd49104dfea4f2606ed1f) as the `style.` and `class.` are dropped. Any ideas on how we can work around this limitation?

Reply:
If you take a look at them on astexplorer.net, e.g. here: https://astexplorer.net/#/gist/077162bbaa73aaf055fefbf24dc54cf3/b3c105a1403d3bb5280ee48a7f3229fa8c507e59

You'll notice that the `BoundAttribute`s have a `type` property on them, and they have different values in each case.

Therefore we will need to dig into what these `type`s map to within `@angular/compiler` and use them to differentiate our logic within the rule

Reply:
(We apply that original `type` from the `@angular/compiler` onto a property called `__originalType` on our final AST so that it is preserved. I don't think we've actually needed it before now!)

Reply:
Ahhh I completely missed that, nice spot! Clicking the definition gives me:
```
export declare const enum BindingType {
    Property = 0,
    Attribute = 1,
    Class = 2,
    Style = 3,
    Animation = 4
}
```

So I can definitely work with that! 😄

Reply:
TIL, really nice to know that. Thanks for sharing this info, @JamesHenry :)